liquidctl/driver/nzxt_smart_device.py

```diff
@@ -182,66 +276,244 @@ def get_status(self, **kwargs):
-            raise NotImplementedError()
-        mval, mod3, mod4, mincolors, maxcolors = _COLOR_MODES[mode]
-        colors = [[g, r, b] for [r, g, b] in colors]
-        if len(colors) < mincolors:
-            raise ValueError('Not enough colors for mode={}, at least {} required'
-                             .format(mode, mincolors))
-        elif maxcolors == 0:
-            if colors:
-                LOGGER.warning('too many colors for mode=%s, none needed', mode)
-            colors = [[0, 0, 0]]  # discard the input but ensure at least one step
-        elif len(colors) > maxcolors:
-            LOGGER.warning('too many colors for mode=%s, dropping to %i',
-                           mode, maxcolors)
-            colors = colors[:maxcolors]
+    def _write_colors(self, cid, mode, colors, sval):
+        mval, mod3, mod4, _, _ = self._COLOR_MODES[mode]
         # generate steps from mode and colors: usually each color set by the user generates
         # one step, where it is specified to all leds and the device handles the animation;
         # but in super mode there is a single step and each color directly controls a led
         if 'super' in mode:
             steps = [list(itertools.chain(*colors))]
         else:
             steps = [color*40 for color in colors]
-        sval = _ANIMATION_SPEEDS[speed]
         for i, leds in enumerate(steps):
             seq = i << 5
             byte4 = sval | seq | mod4
             self._write([0x2, 0x4b, mval, mod3, byte4] + leds[0:57])
             self._write([0x3] + leds[57:])
-        self.device.release()
 
-    def set_fixed_speed(self, channel, duty, **kwargs):
-        """Set channel to a fixed speed."""
-        if channel == 'sync':
-            selected_channels = self._speed_channels
-        else:
-            selected_channels = { channel: self._speed_channels[channel] }
-        for cname, (cid, smin, smax) in selected_channels.items():
-            if duty < smin:
-                duty = smin
-            elif duty > smax:
-                duty = smax
-            LOGGER.info('setting %s duty to %i%%', cname, duty)
-            self._write([0x2, 0x4d, cid, 0, duty])
+    def _write_fixed_duty(self, cid, duty):
+        self._write([0x2, 0x4d, cid, 0, duty])
+
+
+class SmartDeviceDriverV2(CommonSmartDeviceDriver):
+    """liquidctl driver for the NZXT Smart Device V2."""
+
+    SUPPORTED_DEVICES = [
+        (0x1e71, 0x2006, None, 'NZXT Smart Device V2 (experimental)', {
+            'speed_channel_count': 3,
+            'color_channel_count': 2
+        }),
+    ]
+    
+    _READ_LENGTH = 60
+    _WRITE_LENGTH = 64
+
+    _COLOR_MODES = {
+        # (byte2/mode, byte3/variant, byte4/size, min colors, max colors)
+        'off':                              (0x00, 0x00, 0x00, 0, 0),
+        'fixed':                            (0x00, 0x00, 0x00, 1, 1),
+        'super-fixed':                      (0x01, 0x00, 0x00, 1, 40),  # independent leds
```

Comment:
As mentioned above, `super-fixed` is one of the two 40-LED color modes so the same `byte2/mode` value is used as for `fixed`, but we activate this mode with a call to command sequence `\0x22\0xA0` instead of `\0x28\0x03`.

Reply:
Right.  But it's actually sharing `mval` with `fading`, which it is why I though it was odd.

Reply:
Yes, I see...